File: Vantage/Support/RelativeTimeText.swift
import Foundation

/// Renders "Updated 4 minutes ago" strings.
///
/// The spec calls this out specifically: the UI should say how old the data is
/// rather than leaving the user to guess.
///
/// Main-actor isolated because `RelativeDateTimeFormatter` is not `Sendable`
/// and formatter construction is expensive enough that we don't want one per
/// row in a list. Every caller is view code, so this costs nothing in practice.
@MainActor
enum RelativeTimeText {
    private static let formatter: RelativeDateTimeFormatter = {
        let f = RelativeDateTimeFormatter()
        f.unitsStyle = .full
        f.dateTimeStyle = .numeric
        return f
    }()

    static func string(for date: Date, now: Date = .now) -> String {
        // Anything under a few seconds reads oddly as "in 0 seconds".
        if abs(now.timeIntervalSince(date)) < 5 { return "just now" }
        return formatter.localizedString(for: date, relativeTo: now)
    }

    /// The full status line shown under a value, e.g. "Updated 4 minutes ago"
    /// or "Couldn't refresh — showing data from 2 hours ago".
    static func status(for freshness: Freshness, now: Date = .now) -> String {
        switch freshness {
        case .missing:
            "Not available"
        case .fresh(let date):
            "Updated \(string(for: date, now: now))"
        case .stale(let date):
            "Stale — last updated \(string(for: date, now: now))"
        case .refreshing(let previous):
            if let previous {
                "Refreshing — showing data from \(string(for: previous, now: now))"
            } else {
                "Loading…"
            }
        case .failed(let previous, let reason):
            if let previous {
                "Couldn't refresh (\(reason)) — showing data from \(string(for: previous, now: now))"
            } else {
                "Not available (\(reason))"
            }
        }
    }
}

File: core/src/commonMain/kotlin/com/tylerabitbol/libra/support/RelativeTimeText.kt
package com.tylerabitbol.libra.support

import kotlin.math.abs
import kotlin.time.Clock
import kotlin.time.Instant

/**
 * Renders "Updated 4 minutes ago" strings.
 *
 * The spec calls this out specifically: the UI should say how old the data is
 * rather than leaving the user to guess.
 *
 * Swift used `RelativeDateTimeFormatter` with `.full`/`.numeric`, which has no
 * multiplatform equivalent, so the unit selection is written out. It is an
 * approximation of ICU at the month and year boundaries (calendar months vary
 * in length); nothing in the app depends on precision above weeks.
 */
object RelativeTimeText {

    fun string(date: Instant, now: Instant = Clock.System.now()): String {
        val seconds = (now - date).inWholeSeconds
        // Anything under a few seconds reads oddly as "in 0 seconds".
        if (abs(seconds) < 5) return "just now"

        val past = seconds > 0
        val magnitude = abs(seconds)

        val (count, unit) = when {
            magnitude < 60 -> magnitude to "second"
            magnitude < 3_600 -> (magnitude / 60) to "minute"
            magnitude < 86_400 -> (magnitude / 3_600) to "hour"
            magnitude < 604_800 -> (magnitude / 86_400) to "day"
            magnitude < 2_629_746 -> (magnitude / 604_800) to "week"
            magnitude < 31_556_952 -> (magnitude / 2_629_746) to "month"
            else -> (magnitude / 31_556_952) to "year"
        }

        val noun = if (count == 1L) unit else unit + "s"
        return if (past) "$count $noun ago" else "in $count $noun"
    }

    /**
     * The full status line shown under a value, e.g. "Updated 4 minutes ago"
     * or "Couldn't refresh — showing data from 2 hours ago".
     */
    fun status(freshness: Freshness, now: Instant = Clock.System.now()): String =
        when (freshness) {
            is Freshness.Missing -> "Not available"
            is Freshness.Fresh -> "Updated ${string(freshness.asOf, now)}"
            is Freshness.Stale -> "Stale — last updated ${string(freshness.asOf, now)}"
            is Freshness.Refreshing ->
                freshness.previous?.let { "Refreshing — showing data from ${string(it, now)}" }
                    ?: "Loading…"
            is Freshness.Failed ->
                freshness.previous?.let {
                    "Couldn't refresh (${freshness.reason}) — showing data from ${string(it, now)}"
                } ?: "Not available (${freshness.reason})"
        }
}
